doc(backup backing image): add backup and restore backing image doc (#833)
longhorn/website: content/docs/1.6.0/advanced-resources/backing-image-backup.md

---
title: Backing Image Backup
weight: 4
---

As of v1.6.0, Longhorn supports backing up of backing images.

## Prerequisites

> You must first set up a backup target. For more information, see [Setting a Backup Target](../snapshots-and-backups/backup-and-restore/set-backup-target). Without setting up the backup target, Longhorn will report backup target unavailable and won't be able to back up the backing image.

Reviewer: ```suggestion
You must first [set up a backup target](../snapshots-and-backups/backup-and-restore/set-backup-target). If you skip this crucial step, the missing backup target will prevent Longhorn from creating a backup of the backing image.
```

Reviewer: Thanks for describing the error. I summarized the info in my new suggestion.

> ## Create and Restore a Backup of a Backing Image
> 
> Since Backing Image is globally unique in Lnoghorn system, its Backup is also globally unique with the same name.
> 
> You can create a backing image backup by YAML. Here is the example:

Reviewer: ```suggestion
## Create a Backup of a Backing Image

Because backing images are globally unique within the Longhorn system, the corresponding backups are also globally unique and are identified using the same name.

You can create a backups of backing images using YAML. 
```

Reviewer: It seems like this suggestion was missed.

```

You can then create the backup for this backing image by applying the YAML
```yaml
apiVersion: longhorn.io/v1beta2
kind: BackupBackingImage
metadata:
  name: parrot
  namespace: longhorn-system
spec:
  userCreated: true
  labels:
    usecase: test
    type: raw
```

- `name`: should be the same as the backing image or Longhorn won't be able to find the backing image.
- `userCreated`: indicates whether this Backup CR is created by user to create the backup in backupstore or it is synced from backupstrore.
- `labels`: just like volume backup, you can add labels to the backing image backup.

### Restore a backing image from backupstore via YAML

After creating a backup in the backupstore, you are now be able to restore it back in another cluster. Here is the example:

```yaml
apiVersion: longhorn.io/v1beta2
kind: BackingImage
metadata:
    name: parrot-restore
    namespace: longhorn-system
spec:
    sourceType: restore
    sourceParameters:
        # change to your backup URL
        # backup-url: nfs://longhorn-test-nfs-svc.default:/opt/backupstore?backingImage=parrot
        backup-url: s3://backupbucket@us-east-1/?backingImage=parrot

>         concurrent-limit: "2"

Reviewer 2: Where are the credentials?

Author: user don't need to provide the credentials
when Longhorn do the bacuup
it will get the backupTarget and the secret automatically and use that credentials
you can see the code here: https://github.com/longhorn/longhorn-manager/pull/2416/files#diff-3ddec43f26f529baed3c3d660a194662725352c76e4e734d6f7ffbb51325e574R686

its the same as backup volume , user also don't need to provide the credentials to the CR when backup and restore.

    checksum: 304f3ed30ca6878e9056ee6f1b02b328239f0d0c2c1272840998212f9734b196371560b3b939037e4f4c2884ce457c2cbc9f0621f4f5d1ca983983c8cdf8cd9a
```

> - just like creating a backing image except the `soruceType` is now `restore`.
> - required `sourceParameters` when restoring the backing image:
>     - `backup-url`: the url of the backing image resource in the backupstore. You can find this information in the status of the backing image backup CR `.Status.URL`.
>     - `concurrent-limit`: controls how many worker threads per restore concurrently. Will apply default setting if not provided.
> - `checksum`: if provided, Longhorn will validate the restored file with the checksum. If not provided, Longhorn will use the checksum of the restored file as the truth.

Reviewer: ```suggestion
> **IMPORTANT:**
> - `sourceType`: Set the value to `restore`.
> - `sourceParameters`: Configure the following parameters:
>   - `backup-url`: URL of the backing image resource in the backupstore. You can find this information in the status of the backup custom resource `.Status.URL`.
>   - `concurrent-limit`: Maximum number of worker threads that can concurrently run for each restore operation. When unspecified, Longhorn uses the default value.
> - `checksum`: You can specify the expected SHA-512 checksum of the backing image file, which Longhorn uses to validate the restored file. When unspecified, Longhorn uses the checksum of the restored file as the truth.
```

## Volume with a Backing Image

> - When a Volume is backed up, its backing image will be backed up automatically.
> - When restoring a Volume with a backing image
>     - if the backing image already exists in the cluster, Longhorn will use the backing image directly
>     - if the backing image does not exist in the cluster but in the backupstore, Longhorn will restore the backing image automatically.

Reviewer: ```suggestion
When you create a backup of a volume, Longhorn automatically creates a backup of its backing image.

You can restore a volume with a backing image. If the image already exists in the cluster, Longhorn uses the image directly. If the image exists in the backupstore but not in the cluster, Longhorn automatically restores the backing image.
```
